src/Npgsql/NpgsqlConnector.Auth.cs

```diff
@@ -246,58 +253,76 @@ static byte[] Xor(byte[] buffer1, byte[] buffer2)
                 return buffer1;
             }
 
-            static byte[] HMAC(byte[] data, string key)
+            static byte[] HMAC(byte[] key, string data)
             {
-                using var hmacsha256 = new HMACSHA256(data);
-                return hmacsha256.ComputeHash(Encoding.UTF8.GetBytes(key));
+                using var hmacsha256 = new HMACSHA256(key);
+                return hmacsha256.ComputeHash(Encoding.UTF8.GetBytes(data));
             }
         }
 
-        async Task AuthenticateMD5(string username, byte[] salt, bool async, CancellationToken cancellationToken = default)
+        async Task AuthenticateMD5(string username, int salt, bool async, CancellationToken cancellationToken = default)
         {
             var passwd = GetPassword(username);
             if (passwd == null)
                 throw new NpgsqlException("No password has been provided but the backend requires one (in MD5)");
 
-            byte[] result;
+            const int HexHashSize = 2 * 128 / 8;
+
+            var hashLength = HexHashSize + 3 + 1;
+            var hash = ArrayPool<byte>.Shared.Rent(hashLength);
+
             using (var md5 = MD5.Create())
             {
                 // First phase
-                var passwordBytes = PGUtil.UTF8Encoding.GetBytes(passwd);
-                var usernameBytes = PGUtil.UTF8Encoding.GetBytes(username);
-                var cryptBuf = new byte[passwordBytes.Length + usernameBytes.Length];
-                passwordBytes.CopyTo(cryptBuf, 0);
-                usernameBytes.CopyTo(cryptBuf, passwordBytes.Length);
+                var passwdLength = PGUtil.UTF8Encoding.GetByteCount(passwd);
+                var usernameLength = PGUtil.UTF8Encoding.GetByteCount(username);
+
+                var credentialsLength = passwdLength + usernameLength;
+                var credentials = ArrayPool<byte>.Shared.Rent(credentialsLength);
+
+                var preHashLength = HexHashSize + 4;
+                var preHash = ArrayPool<byte>.Shared.Rent(preHashLength);
 
-                var sb = new StringBuilder();
-                var hashResult = md5.ComputeHash(cryptBuf);
-                foreach (var b in hashResult)
-                    sb.Append(b.ToString("x2"));
+                PGUtil.UTF8Encoding.GetBytes(passwd, 0, passwdLength, credentials, 0);
+                PGUtil.UTF8Encoding.GetBytes(username, 0, usernameLength, credentials, passwdLength);
 
-                var prehash = sb.ToString();
+                HexHash(preHash, credentials, credentialsLength);
+                Unsafe.WriteUnaligned(ref preHash[HexHashSize], salt);
 
-                var prehashbytes = PGUtil.UTF8Encoding.GetBytes(prehash);
-                cryptBuf = new byte[prehashbytes.Length + 4];
+                // Second phase
+                hash[0] = (byte)'m';
+                hash[1] = (byte)'d';
+                hash[2] = (byte)'5';
+                hash[hashLength - 1] = 0;
```

Comment:
Micro-optimization hint:
Here the JIT need to emit a bound check for each access. If the order is swapped, so the largest index access comes first, the JIT just needs to emit one bound check and can elide the other ones.